src/main/kotlin/com/exactpro/th2/codec/csv/CsvCodec.kt

```diff
@@ -0,0 +1,318 @@
+                continue
+            }
+            val rawMessage = anyMessage.rawMessage
+            val protocol = rawMessage.metadata.protocol
+            if ("" != protocol && !"csv".equals(protocol, ignoreCase = true)) {
+                LOGGER.error { "Wrong protocol: message should have empty or 'csv' protocol but has $protocol" }
+                continue
+            }
+            val data = decodeValues(rawMessage.body.toByteArray())
+            if (data.isEmpty()) {
+                LOGGER.error { "The raw message does not contains any data: ${rawMessage.toJson()}" }
+                errors.add(ErrorHolder("The raw message does not contains any data", rawMessage))
+                continue
+            }
+            decodeCsvData(errors, groupBuilder, rawMessage, data)
+        }
+        if (errors.isNotEmpty()) {
+            throw DecodeException(
+                "Cannot decode some messages:\n" + errors.joinToString("\n") {
+                    "Message ${it.originalMessage.logId} cannot be decoded because ${it.text}"
+                }
+            )
+        }
+        return groupBuilder.build()
+    }
+
+    override fun decode(messageGroup: MessageGroup, context: IReportingContext): MessageGroup = decode(messageGroup)
+    override fun decode(messageGroup: MessageGroup): MessageGroup {
+        val decodedMessages = mutableListOf<Message<*>>()
+        val errors: MutableCollection<ErrorHolder<RawMessage>> = mutableListOf()
+
+        for (message in messageGroup.messages) {
+            if (message is ParsedMessage) {
+                decodedMessages.add(message)
+                continue
+            }
+
+            if (message !is RawMessage) {
+                LOGGER.error { "Message should either have a raw or parsed message but has nothing: $message" }
+                continue
+            }
+
+            val protocol = message.protocol
+            if ("" != protocol && !"csv".equals(protocol, ignoreCase = true)) {
+                LOGGER.error { "Wrong protocol: message should have empty or 'csv' protocol but has $protocol" }
+                continue
+            }
+
+            val data = decodeValues(message.body.toByteArray())
+
+            if (data.isEmpty()) {
+                LOGGER.error { "The raw message does not contains any data: ${message.toJson()}" }
+                errors.add(ErrorHolder<RawMessage>("The raw message does not contains any data", message))
+                continue
+            }
+
+            decodeCsvData(errors, decodedMessages, message, data)
+        }
+
+        if (!errors.isEmpty()) {
+            throw DecodeException(
+                "Cannot decode some messages:\n$" + errors.joinToString("\n") {
+                    "Message ${it.originalMessage.toJson()} cannot be decoded because ${it.text}"
+                }
+            )
+        }
+
+        return MessageGroup(decodedMessages)
+    }
+
+    private fun decodeCsvData(
```

Comment:
Could we extract some part of this method to a separate method to exclude copy past?